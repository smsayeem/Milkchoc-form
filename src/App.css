
/* google font */
@import url('https://fonts.googleapis.com/css2?family=Tinos:wght@400;700&display=swap');

/* color variables */
:root {
  --primaryGrey: rgba(61,79,97, 0.8);
  --primaryLight: rgba(61,79,97, 0.4);

  --primary: rgb(93,168,221);
  --primaryDark: rgb(30, 112, 172);

  --mainWhite: rgb(255,255,255);
  --mainGrey: rgb(231,233,236);
  --mainBlack: rgb(50,50,50);

  --error:rgb(255,0,0);
  --errorLight:rgb(255,243,243)
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
}

body {
  background: var(--mainBlack);
  color: var(--primaryColor);
  font-family: 'Tinos', serif;
}
