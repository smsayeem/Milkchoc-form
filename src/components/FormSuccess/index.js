import React from 'react';
import styled from 'styled-components';

export default function FormSuccess() {
  return (
    <StyledContainer>
      <h1>We have received your request!</h1>
      <StyledButton href='/'>Go back</StyledButton>
    </StyledContainer>
  )
}

const StyledButton = styled.a`
  text-decoration: none;
  border: 1px solid var(--primaryGrey);
  color: var(--primaryDark);
  display: block;
  border-radius: 4px;
  margin: 2em;
  text-align: center;

  &:hover {
    cursor: pointer;
    color: var(--primary);
    background-color: var(--mainGrey);
  }
`;

const StyledContainer = styled.div`
  font-size: 2rem;
  color: var(--primaryDark);
  margin: 0 auto;
`;
