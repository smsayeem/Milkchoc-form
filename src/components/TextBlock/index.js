import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

function TextBlock({as, text}) {
  switch (as) {
    case 'h1':
      return <H1>{text}</H1>
    case 'h2':
      return <H2>{text}</H2>
    case 'p':
    default:
      return <P>{text}</P>
  }
}

const H1 = styled.h1`
  letter-spacing: 0.8px;
  font-size: 1.6rem;
  margin-bottom: 1.5rem;

  @media screen and (min-width: 769px){
    letter-spacing: 1px;
    font-size: 2rem;
    margin-bottom: 2.5rem;
    }
`;
const H2 = styled.h2`
  letter-spacing: 0.8px;
  font-size: 1.2rem;
  margin-bottom: 1.5rem;

  @media screen and (min-width: 769px){
    letter-spacing: 1px;
    font-size: 1.5rem;
    margin-bottom: 1.5rem;
    }
`;
const P = styled.p``;

TextBlock.propTypes = {
  as: PropTypes.string, 
  text: PropTypes.string.isRequired
};

TextBlock.defaultProps = {
  as: 'p'
};

export default TextBlock
