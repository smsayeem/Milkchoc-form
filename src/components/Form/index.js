import React, { useState } from 'react';
import PropTypes from 'prop-types';
import validateInfo from './components/FormValidation/validateInfo';
import styled, {css} from 'styled-components';
import TextBlock from '../TextBlock';

function Form({handleChange, info, handleSubmit} ) {
  const [err, setErr] = useState({})

  const validateFormData = (e) => {
    e.preventDefault();
    const myErr = validateInfo(info);
    setErr(myErr);

    if(Object.keys(myErr).length === 0){
      handleSubmit(e);
    }
  }

  return (
    <FormContainer>
      <form onSubmit={validateFormData}>
        <TextBlock as='h1' text='Fill out this awesome form' />
        <FormCenter>
          <TextBlock as='h2' text='Your details' />
          <FormGroup error={err.email}>
            <StyledLabel htmlFor="email">Email</StyledLabel>
            <InputWrapper>
              <StyledInput 
                type="email" 
                id="email"
                name='email'
                placeholder="Enter your email address"
                value={info.email}
                className="input-box"
                onChange={handleChange}  />
              {err.email && <TextBlock as='p' text={err.email} />}
            </InputWrapper>
          </FormGroup>
          <FormGroup error={err.password}>
            <StyledLabel htmlFor="password">Password</StyledLabel>
            <InputWrapper>
              <StyledInput 
                type="password" 
                id="password"
                name='password'
                placeholder="Choose your password"
                value={info.password}
                className="input-box"
                onChange={handleChange}  />
              {err.password && <TextBlock as='p' text={err.password} />}
            </InputWrapper>
          </FormGroup>
          <Divider />
          <TextBlock as='h2' text='Your animal' />
          <FormGroup error={err.color}>
            <StyledLabel htmlFor="color">color</StyledLabel>
            <InputWrapper>
              <StyledSelect name="color" id="color" className="input-box" onChange={handleChange}>
                <option value="">Choose your color</option>
                <option value="blue">Blue</option>
                <option value="green">Green</option>
                <option value="red">Red</option>
                <option value="black">Black</option>
                <option value="brown">Brown</option>
              </StyledSelect>
              {err.color && <TextBlock as='p' text={err.color} />}
            </InputWrapper>
          </FormGroup>
          <FormGroup error={err.animal}>
            <StyledLabel>Animal</StyledLabel>
            <CheckboxWrapper>
              <div>
                <StyledCheckbox 
                  type='checkbox'
                  name='animal'
                  value='Bear'
                  className='chkbox-input'
                  onChange={handleChange} />
                <CheckboxLabel>Bear</CheckboxLabel>
              </div>
              <div>
                <StyledCheckbox 
                  type='checkbox'
                  name='animal'
                  value='tiger'
                  className='chkbox-input'
                  onChange={handleChange} />
                <CheckboxLabel>Tiger</CheckboxLabel>
              </div>
              <div>
                <StyledCheckbox 
                  type='checkbox'
                  name='animal'
                  value='Snake'
                  className='chkbox-input'
                  onChange={handleChange} />
                <CheckboxLabel>Snake</CheckboxLabel>
              </div>
              <div>
                <StyledCheckbox 
                  type='checkbox'
                  name='animal'
                  value='Donkey'
                  className='chkbox-input'
                  onChange={handleChange} />
                <CheckboxLabel>Donkey</CheckboxLabel>
              </div>
              {err.animal && <TextBlock as='p' text={err.animal} />}
            </CheckboxWrapper>
          </FormGroup>
          <FormGroup error={err.tigertype}>
            {info.animal.map(item => item === 'tiger' ?
            <span key={item}>
              <StyledLabel htmlFor="tigertype">Tiger type</StyledLabel>
              <InputWrapper>
                <StyledInput 
                  type="text"
                  id='tigertype'
                  name="tigertype"
                  placeholder="Enater tiger type"
                  className="input-box"
                  value={info.tigertype}
                  onChange={handleChange}
                />
                {err.tigertype && <TextBlock as='p' text={err.tigertype} />}
              </InputWrapper>
            </span>
            : 
            null
            )}
            </FormGroup>
          <Divider />
          <ButtonWrapper>
            <StyledButton type="submit" name="submit">Create account</StyledButton>
          </ButtonWrapper>
        </FormCenter>
      </form>
    </FormContainer>
  )
}

const FormContainer = styled.div`
    display: flex;
    flex-direction: column;
    padding: 3rem 1rem 1rem;
    background-color: var(--mainWhite);
    max-width: 640px;
    width: 100%;
    margin: 0 auto;

    @media screen and (min-width: 769px){
      padding: 3rem;
    }
  `;

const FormCenter = styled.section`
  display: flex;
  flex-direction: column;
  margin-top: 24px;
`;

const StyledLabel = styled.label`
  width: 25%;
  color: var(--primaryGrey);
  text-transform: capitalize;
  font-size: 1.1rem;

  @media screen and (min-width: 769px){
    width: 30%;
    }
`;

const InputWrapper = styled.div`
  width: 75%;

  @media screen and (min-width: 769px){
    width: 70%;
    }
`;

const StyledInput = styled.input`
  width: 100%;
  border: 1px solid var(--primaryLight);
  border-radius: 4px;
  outline: none;
  font-size: 1rem;
  padding: 10px;
  color: var(--mainBlack);
  letter-spacing: 1px;

  &[type='text']:focus,
  &[type='email']:focus,
  &[type='password']:focus{
    outline: none;
    border-color: var(--primary);
    box-shadow: 0 0 6px rgba(48, 140, 255, 0.25);
  }
  &::placeholder{
    color: var(--primaryLight);
    font-size: .9rem;
  }
`;

const StyledSelect = styled.select`
  -webkit-appearance: none;
  width: 100%;
  border: 1px solid var(--primaryLight);
  border-radius: 4px;
  outline: none;
  font-size: 1rem;
  padding: 10px;
  color: var(--mainBlack);
  letter-spacing: 1px;

`;

const CheckboxWrapper = styled.div`
  display: flex;
  align-items: center;
  width: 70%;
  justify-content: space-between;
  flex-wrap: wrap;
`;

const CheckboxItem = styled.div`

`;

const StyledCheckbox = styled.input`
  margin: 0 8px;
`;
const CheckboxLabel = styled.label`
  padding: 10px 0;
  color: var(--primaryGrey);
`;
const FormGroup = styled.div`
  display: flex;
  align-items: center;
  margin: 10px 0;

  ${props => props.error && css`
    & input{
      color:  var(--error);
      background-color: var(--errorLight);
    }
    &, & > label{
      color:  var(--error);
    }
  `}
`;
const ButtonWrapper = styled.div`
  padding: 1rem 0;
  width: 100%;
  display: flex;
  justify-content: space-between;

  @media screen and (min-width: 769px){
    margin: 1rem 0 0 0;
    }
`;

const StyledButton = styled.button`
  width: 48%;
  height: 2.7rem;
  text-decoration: none;
  color: var(--mainWhite);
  background-color: var(--primary);
  text-align: center;
  text-transform: capitalize;
  letter-spacing: 1px;
  font-size: 1rem;
  outline: 0;
  border: none;
  border-radius: 4px;
  padding: 0 1rem;
  cursor: pointer;
  box-shadow: 0 2px 2px rgba(0, 0, 0, 0.2);
  transition: all 0.3s ease;

  &:hover, &:active{
    background-color: var(--primaryDark);
    box-shadow: 0 3px 3px rgba(0, 0, 0, 0.3); 
  }

  @media screen and (min-width: 769px){
    font-size: 1.1rem;
    }
`;

const Divider = styled.div`
  border-bottom: 1px solid var(--mainGrey);
  margin: 1rem 0;
`;

Form.propTypes = {
  handleChange: PropTypes.func, 
  handleSubmit: PropTypes.func, 
  info: PropTypes.shape({
    email: PropTypes.string,
    password: PropTypes.string,
    color: PropTypes.string,
    animal: PropTypes.array,
    tigertype:PropTypes.string
  }),
};

Form.defaultProps = {
  handleChange: ()=>{}, 
  handleSubmit: ()=>{}, 
  info: {
    email: '',
    password: '',
    color: '',
    animal: [],
    tigertype:''
  }
};

export default Form
