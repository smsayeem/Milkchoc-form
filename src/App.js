import React, { useState } from 'react';
import Form from './components/Form';
import FormSuccess from './components/FormSuccess';
import './App.css';
import styled from 'styled-components';

function App() {

    const [info, setInfo] = useState({
      email:'',
      password:'',
      color:'',
      animal:[],
      tigertype:''
    })
    const [submit, setSubmit] = useState(false);

    const handleChange = (e) => {
      const {name, value} = e.target;
      if(e.target.type === 'checkbox'){

        if(e.target.checked){
          setInfo({...info, [name]:[...info.animal, value]});
        }else{
          if(value === 'tiger'){
            info.tigertype = '';
          }
          const newAnimal = info.animal.filter(item=> item !== value);
          setInfo({...info, [name]:newAnimal});
        }

      }else{
        setInfo({ ...info, [name]:value });
      }

    }
  
    // No functionality added for the submit button.
    const handleSubmit = (e) => {
      console.log('state success=',info);
      console.log('congratulation !!! Form submitted successfully.');
      setSubmit(true);
    }

  return (
    <StyledApp>
      {!submit ? 
        (<Form 
          handleChange={handleChange}
          handleSubmit={handleSubmit} 
          info={info}
        />)
      : 
        (<FormSuccess />)
      }
    </StyledApp>
  );
}

const StyledApp = styled.div`
  display: flex;
  align-items: flex-start;
  background-color: var(--mainGrey);
  min-height: 100vh;
`;

export default App;
