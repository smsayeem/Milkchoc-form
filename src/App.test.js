import { shallow, mount } from 'enzyme';
import React from 'react';
import App from './App';
import Form from './components/Form';
import FormSuccess from './components/FormSuccess'
import TextBlock from './components/TextBlock';

// components existance check
describe('App and other components existence check', () => {
  // let wrapper;
  //   beforeEach(()=>{
  //     wrapper = shallow(<App />);
  // });

  // snapshot test
  it('should render the <App /> Component correctly', () => {  
    const wrapper = mount(<App />); 
    expect(wrapper).toMatchSnapshot();
  });

  // render the App componnet without crashing
  it('should render the App componnet without crashing', ()=>{
    shallow(<App />);
  });
  // render the Form componnet without crashing
  it('should render the Form componnet without crashing', ()=>{
    shallow(<Form />);
  });
      // render the TextBlock componnet without crashing
  it('should render the TextBlock componnet without crashing', ()=>{
    shallow(<TextBlock as="h1" text="Test text" />);
  });

  it('should render the FormSuccess componnet without crashing', ()=>{
    shallow(<FormSuccess />);
  });
});
