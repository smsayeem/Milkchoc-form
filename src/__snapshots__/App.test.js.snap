// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App and other components existence check should render the <App /> Component correctly 1`] = `
<App>
  <styled.div>
    <div
      className="sc-fFubgz gzKReG"
    >
      <HelmetWrapper
        defer={true}
        encodeSpecialCharacters={true}
      >
        <SideEffect(NullComponent)
          defer={true}
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "Milkchoc form app is a demo form, for SEO purpose",
                "name": "description",
              },
            ]
          }
          title="Milkchoc form"
          titleAttributes={Object {}}
        >
          <NullComponent
            defer={true}
            encodeSpecialCharacters={true}
            meta={
              Array [
                Object {
                  "content": "Milkchoc form app is a demo form, for SEO purpose",
                  "name": "description",
                },
              ]
            }
            title="Milkchoc form"
            titleAttributes={Object {}}
          />
        </SideEffect(NullComponent)>
      </HelmetWrapper>
      <Form
        handleChange={[Function]}
        handleSubmit={[Function]}
        info={
          Object {
            "animal": Array [],
            "color": "",
            "email": "",
            "password": "",
            "tigertype": "",
          }
        }
      >
        <styled.div>
          <div
            className="sc-hKgILt kThEky"
          >
            <form
              onSubmit={[Function]}
            >
              <TextBlock
                as="h1"
                text="Fill out this awesome form"
              >
                <styled.h1>
                  <h1
                    className="sc-bdfBwQ kEmZII"
                  >
                    Fill out this awesome form
                  </h1>
                </styled.h1>
              </TextBlock>
              <styled.section>
                <section
                  className="sc-eCssSg ggsmVZ"
                >
                  <TextBlock
                    as="h2"
                    text="Your details"
                  >
                    <styled.h2>
                      <h2
                        className="sc-gsTCUz drecMv"
                      >
                        Your details
                      </h2>
                    </styled.h2>
                  </TextBlock>
                  <styled.div>
                    <div
                      className="sc-crrsfI lnWkZO"
                    >
                      <styled.label
                        htmlFor="email"
                      >
                        <label
                          className="sc-jSgupP boUXXQ"
                          htmlFor="email"
                        >
                          Email
                        </label>
                      </styled.label>
                      <styled.div>
                        <div
                          className="sc-gKsewC jHGHpE"
                        >
                          <styled.input
                            className="input-box"
                            id="email"
                            name="email"
                            onChange={[Function]}
                            placeholder="Enter your email address"
                            type="email"
                            value=""
                          >
                            <input
                              className="sc-iBPRYJ hRCnUG input-box"
                              id="email"
                              name="email"
                              onChange={[Function]}
                              placeholder="Enter your email address"
                              type="email"
                              value=""
                            />
                          </styled.input>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-crrsfI lnWkZO"
                    >
                      <styled.label
                        htmlFor="password"
                      >
                        <label
                          className="sc-jSgupP boUXXQ"
                          htmlFor="password"
                        >
                          Password
                        </label>
                      </styled.label>
                      <styled.div>
                        <div
                          className="sc-gKsewC jHGHpE"
                        >
                          <styled.input
                            className="input-box"
                            id="password"
                            name="password"
                            onChange={[Function]}
                            placeholder="Choose your password"
                            type="password"
                            value=""
                          >
                            <input
                              className="sc-iBPRYJ hRCnUG input-box"
                              id="password"
                              name="password"
                              onChange={[Function]}
                              placeholder="Choose your password"
                              type="password"
                              value=""
                            />
                          </styled.input>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-kstrdz dIMTXz"
                    />
                  </styled.div>
                  <TextBlock
                    as="h2"
                    text="Your animal"
                  >
                    <styled.h2>
                      <h2
                        className="sc-gsTCUz drecMv"
                      >
                        Your animal
                      </h2>
                    </styled.h2>
                  </TextBlock>
                  <styled.div>
                    <div
                      className="sc-crrsfI lnWkZO"
                    >
                      <styled.label
                        htmlFor="color"
                      >
                        <label
                          className="sc-jSgupP boUXXQ"
                          htmlFor="color"
                        >
                          color
                        </label>
                      </styled.label>
                      <styled.div>
                        <div
                          className="sc-gKsewC jHGHpE"
                        >
                          <styled.select
                            className="input-box"
                            id="color"
                            name="color"
                            onChange={[Function]}
                          >
                            <select
                              className="sc-fubCfw daNcGr input-box"
                              id="color"
                              name="color"
                              onChange={[Function]}
                            >
                              <option
                                value=""
                              >
                                Choose your color
                              </option>
                              <option
                                value="blue"
                              >
                                Blue
                              </option>
                              <option
                                value="green"
                              >
                                Green
                              </option>
                              <option
                                value="red"
                              >
                                Red
                              </option>
                              <option
                                value="black"
                              >
                                Black
                              </option>
                              <option
                                value="brown"
                              >
                                Brown
                              </option>
                            </select>
                          </styled.select>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-crrsfI lnWkZO"
                    >
                      <styled.label>
                        <label
                          className="sc-jSgupP boUXXQ"
                        >
                          Animal
                        </label>
                      </styled.label>
                      <styled.div>
                        <div
                          className="sc-pFZIQ gEyQsu"
                        >
                          <div>
                            <styled.input
                              className="chkbox-input"
                              name="animal"
                              onChange={[Function]}
                              type="checkbox"
                              value="Bear"
                            >
                              <input
                                className="sc-kEjbxe uUWTw chkbox-input"
                                name="animal"
                                onChange={[Function]}
                                type="checkbox"
                                value="Bear"
                              />
                            </styled.input>
                            <styled.label>
                              <label
                                className="sc-iqHYGH dhtXYU"
                              >
                                Bear
                              </label>
                            </styled.label>
                          </div>
                          <div>
                            <styled.input
                              className="chkbox-input"
                              name="animal"
                              onChange={[Function]}
                              type="checkbox"
                              value="tiger"
                            >
                              <input
                                className="sc-kEjbxe uUWTw chkbox-input"
                                name="animal"
                                onChange={[Function]}
                                type="checkbox"
                                value="tiger"
                              />
                            </styled.input>
                            <styled.label>
                              <label
                                className="sc-iqHYGH dhtXYU"
                              >
                                Tiger
                              </label>
                            </styled.label>
                          </div>
                          <div>
                            <styled.input
                              className="chkbox-input"
                              name="animal"
                              onChange={[Function]}
                              type="checkbox"
                              value="Snake"
                            >
                              <input
                                className="sc-kEjbxe uUWTw chkbox-input"
                                name="animal"
                                onChange={[Function]}
                                type="checkbox"
                                value="Snake"
                              />
                            </styled.input>
                            <styled.label>
                              <label
                                className="sc-iqHYGH dhtXYU"
                              >
                                Snake
                              </label>
                            </styled.label>
                          </div>
                          <div>
                            <styled.input
                              className="chkbox-input"
                              name="animal"
                              onChange={[Function]}
                              type="checkbox"
                              value="Donkey"
                            >
                              <input
                                className="sc-kEjbxe uUWTw chkbox-input"
                                name="animal"
                                onChange={[Function]}
                                type="checkbox"
                                value="Donkey"
                              />
                            </styled.input>
                            <styled.label>
                              <label
                                className="sc-iqHYGH dhtXYU"
                              >
                                Donkey
                              </label>
                            </styled.label>
                          </div>
                        </div>
                      </styled.div>
                    </div>
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-crrsfI lnWkZO"
                    />
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-kstrdz dIMTXz"
                    />
                  </styled.div>
                  <styled.div>
                    <div
                      className="sc-dQppl iuotMY"
                    >
                      <styled.button
                        name="submit"
                        type="submit"
                      >
                        <button
                          className="sc-bqyKva iIMsjd"
                          name="submit"
                          type="submit"
                        >
                          Create account
                        </button>
                      </styled.button>
                    </div>
                  </styled.div>
                </section>
              </styled.section>
            </form>
          </div>
        </styled.div>
      </Form>
    </div>
  </styled.div>
</App>
`;
